Optimize kafka format

Dieterbe opened the pull request:
see #199 

adds support for the following payloads incoming from kafka:
(codenames, will be changed later if we take this further)
```
// useful for singletenant setups where we can just configure orgid
type MetricPointId1 struct {
        Id    [16]byte
        Value float64
        Time  uint32
}

// for multi-tenant setups, need to pass orgid in the payloads
type MetricPointId2 struct {
        MetricPointId1
        Org uint32
}

```
(in the actual messages they are preceded by a 1byte format header)
because we now encode the id's in a more optimized form, i will also attempt to update the codebase itself to work with these id's instead of strings; in particular:

```
// when the orgid is implied already
type Key [16]byte

// multinenant
type MKey struct {
        Key [16]byte
        Org uint32
}
```

Dieterbe commented:
## goals and consequences -> extra sub-goals
1) smaller kafka mdm messages, less allocation/cpu/memory heavy serializing and deserializing of the messages
2) because of the more optimized format, it now also makes sense to update all our metric identifiers internally to be based on byte arrays rather than strings, in the cache, AggMetrics, index, etc. (less GC load, takes less space, should be a bit cpu friendlier also)
3) because our internal index format uses the MetricDefinition type, we update the id of the type itself (otherwise we pay the cost of frequent conversions as shown in #848)
(should be harmless as explained in the commit message).

it would be prettier to do these steps one at the time but because they're so tangled let's just bite the bullet and do them at once. also helps that we won't have to test again later.

## status
Making good progress on this. main things left to do:
* make index support both incoming types in AddOrUpdate. likely via the new msg.Point format
* make chunk cache work again
* update all `Req` related things (incoming http requests, intra-cluster requests, etc) to use the new id/key types.
* real-world benchmarking / load testing

## what is changing in this PR:
* new data payload formats and Key types in schema
* kafka plugin to support the new format
* index needs to support old and new types in AddOrCreate call. currently looking for it to take in the dual-type msg.Point, as opposed to creating separate methods for each type
* anywhere we use metric id's (aggmetric/aggmetrics/aggregator/notifier/CWR/store/...) now work with the new keys. nowhere in the code shall metric id's be strings other than when converting to cassandra rows or columns or notifier messages, or reading the old format.
* internal request types for the http api, intra-cluster traffic

## what is NOT changing:
* MetricData or MetricDataArray types we don't need to change it. we can keep using it (sparingly) for now, a new method to submit "full" payloads (with definition) can be done later but probably low priority.
* other input plugins
* persist notifier message format (minimal conversion code)
* chunk data/schema in cassandra (minimal conversions to get the string id's)
* index data in cassandra  (minimal conversions to get the string id's)


## note to self. to test:
* data saved to correct place, both raw and rollup
* reads look good, both raw and rollup
* clustered index calls
* index saving/loading

## deployment
* writers can be upgraded in place.
* clustered readers cannot: will need to create a new reader cluster then destroy old one (e.g. via the color mechanism in hosted-metrics-api)  too much has changed. in particular:
 - intra cluster traffic cluster nodes talking to each other (Req types and MetricDefinition type)


## to reviewers
* commit order in github is messed up due to frequent history rewriting. to see history, check out locally and use https://github.com/jonas/tig or similar. the commits try to tell a story, you'll see the schema changes and then the metrictank changes they enable.
looking at the schema stuff first is probably best to then understand how it's used in metrictank.
* one of my favorite side effects of this PR is that uses of the key are now much more explicit.
e.g. where do we use regular keys like 123.deadbeef vs 123.deadbeef_avg_600 for aggregates, in the past this was often hard to follow, with the new code the new key types make this explicit.

woodsaj commented on idx/cassandra/cassandra.go:
```diff
@@ -365,8 +365,15 @@ func (c *CasIdx) load(defs []schema.MetricDefinition, iter cqlIterator, cutoff u
 	cutoff64 := int64(cutoff)
 	for iter.Scan(&id, &orgId, &partition, &name, &metric, &interval, &unit, &mtype, &tags, &lastupdate) {
+		mkey, err := schema.MKeyFromString(id)
+		if er != nil {
```
should be `err` not `er`

woodsaj commented on msg/msg.go:
```diff
@@ -0,0 +1,11 @@
+// Point represents an incoming datapoint
+// represented either as MetricData or MetricPointId2
+type Point struct {
+	Val   uint8 // 0 means use md, 1 means use point
```
`Val` seems like the wrong term to use here. Perhaps we use `DataType` and then use
```
var MetricData = uint8(0)
var MetricPoint = uint8(1)
```
Which will make the code much more readable
```
p := msg.Point{
   DataType: msg.MetricPoint,
   Point: point,
}
```

Dieterbe replied:
yep. or const even not var. but i decided to do away with this composite msg.Point type and instead use different codepaths in the input handler and index. will push soon.

replay commented on mdata/cache/ccache.go:
```diff
@@ -125,20 +126,20 @@ func (c *CCache) CacheIfHot(metric string, prev uint32, itergen chunk.IterGen) {
 }
 
-func (c *CCache) Add(metric, rawMetric string, prev uint32, itergen chunk.IterGen) {
+func (c *CCache) Add(metric, rawMetric schema.AMKey, prev uint32, itergen chunk.IterGen) {
```
Does this still require both params `metric` and `rawMetric`? I think just one `schema.AMKey` contains all the information that's needed, because the raw key can be extracted from it.

Dieterbe replied:
you're right

replay commented on mdata/cache/ccache.go:
```diff
@@ -31,10 +32,10 @@ type CCache struct {
 
 	// sets of metric keys, indexed by their raw metric keys
-	metricRawKeys map[string]map[string]struct{}
+	metricRawKeys map[schema.AMKey]map[schema.AMKey]struct{}
```
could this be optimized into something like `map[schema.AMKey]map[schema.Archive]`?

replay commented on mdata/cache/ccache_metric.go:
```diff
@@ -20,7 +21,7 @@ type CCacheMetric struct {
 	keys []uint32
 
-	SuffixLen uint8
+	rawKey schema.AMKey
```
shouldn't `schema.MKey` be sufficient here?

woodsaj commented on input/kafkamdm/kafkamdm.go:
```diff
@@ -349,15 +351,30 @@ func (k *KafkaMdm) consumePartition(topic string, partition int32, currentOffset
-	md := schema.MetricData{}
-	_, err := md.UnmarshalMsg(data)
+	pointMsg := msg.Point{}
+	if len(data) == 29 && data[0] == 0 {
```
I think that the unmarshaling of bytes to a msg.Point should be handled in the `msg` package.

Dieterbe replied:
moved it

woodsaj commented on mdata/notifierKafka/notifierKafka.go:
```diff
@@ -232,7 +234,13 @@ func (c *NotifierKafka) flush() {
 	var pMsg mdata.PersistMessageBatch
 	for i, msg := range c.buf {
-		def, ok := c.idx.Get(strings.SplitN(msg.Key, "_", 2)[0])
+		mkey, err := schema.MKeyFromString(strings.SplitN(msg.Key, "_", 2)[0])
```
shouldn't msg.Key be an AMKey?

Dieterbe replied:
msg.Key is SavedChunk.Key which is a string because I don't want to change the notifier format at this point. it does represent a schema.AMKey (see SavedChunk definition)

if your point is that it's a bit confusing that we're parsing out an MKey here, then i agree. I actually attempted to write an AMKeyFromString function but skipped it for now. maybe i can give it another shot.
but anyway we only need an MKey for the index lookup so doing the string.Split and then parsing that seemed ok to me.

Dieterbe replied:
I cleaned this up today.

Dieterbe commented on vendor/gopkg.in/raintank/schema.v1/metric.go:
```diff
@@ -32,6 +32,91 @@ type PartitionedMetric interface {
 
 //go:generate msgp
 
+type MetricPointId1 struct {
```
> I really dont think we need two MetricPointId types. If we do, they need meaningful names.

i agree. i wanted to use the name MetricPoint but you used that in your pr so i had to separate them at first so i could compare the different implementations.
as i said in OP they are codenames that will be renamed.

i like your idea of switching to a single type though, that will make the code easier.
also while we're at it, I want to use the MKey type in this struct.

shanson7 commented on vendor/gopkg.in/raintank/schema.v1/metricpoint.go:
```diff
@@ -0,0 +1,93 @@
+	return bts[28:], nil
+}
+
+// MMarshal32 marshals the MetricPoint directly to b.
```
"Marshal32"

Dieterbe commented:
note: since the idx structures are mostly already by (or could be restructured to be by) orgid we could potentially use Keys instead of MKey's for e.g IdSet, defById. that would save some memory. future optimization, not going to bother with it now I think.

Dieterbe commented on idx/cassandra/cassandra.go:
```diff
@@ -264,10 +264,36 @@ func (c *CasIdx) Stop() {
+	// abeit very unlikely,
+	// the idx entry could be pruned in between the two calls and so they could be different
+	existing, inMemory := c.MemoryIdx.Get(point.MKey)
+	archive, inMemory2 := c.MemoryIdx.UpdateMaybe(point, partition)
```
> Is it possible that UpdateMaybe could return existing as well?

it does, but after it updates the LastUpdate and Partition fields.
that said, the only thing we really need `existing` for is the existing.Partition attribute which determines if we have to clean up the old entry. UpdateMaybe could just return that value. Not very elegant but could be worthwhile.

shanson7 replied:
Right, I meant `existing` as it is in this snippet (i.e. the value before update). If you don't foresee another need to pop up other than the partition value, then that would likely be the most efficient solution and could greatly reduce the contention on this mutex (conjecture).

Dieterbe replied:
i'll give it a shot :)

Dieterbe replied:
this is done now btw

Dieterbe commented on idx/cassandra/cassandra.go:
```diff
@@ -264,10 +264,36 @@ func (c *CasIdx) Stop() {
+	}
+
+	if inMemory2 {
+		archive = c.updateCassandraIfStale(inMemory, archive, partition)
```
> One example we have is that we don't include the Interval in the id we generate (for reasons I can explain if you are curious).

always :)

I think I don't really understand the question/problem,
but there's 2 places where this logic is called (UpdateMaybe which takes the MetricPoint, and AddOrUpdate which takes a full MetricData). however, by the time we get to the updateCassandraIfStale call, it doesn't really matter which of the 2 paths we took to get to this logic, as the calls use the same parameters which it pulled out of the memory index (possibly after an update), and the partition value.
In that regard, whether you sent a MetricData (with the same interval) or a MetricPoint, the logic should be equivalent.
and the ordering should remain honored throughout the kafka->CassandraIndex->Cassandra pipeline. 

does this help @shanson7 ? otherwise let's discuss further

shanson7 replied:
My original comment can be ignored. As long as any data that is not encoded into the `Id` is in the MetricPoint, then this should be sufficient.

Dieterbe replied:
there is one field from MetricData that doesn't contribute to its Id, which is `Name`. However, since we always say that field should equal the `Metric` field (which does contribute the id), this should be OK. Note that because of this, we plan to remove one of these two as they should always be the same.

shanson7 replied:
Yes, I noticed that. I use `Name` when encoding and set `Metric` to `.` for everything (to reduce the size of the messages)

Dieterbe commented:
making progress. have been able to do some useful tests
here's a run that shows old format, new, old, new. all using snappy compression with kafka

https://snapshot.raintank.io/dashboard/snapshot/ZS91Vehxxsd3xp5RTtXeyPu5txOhdbaJ?orgId=2
see https://snapshot.raintank.io/dashboard/snapshot/P14i6y2q55sYjluu5q9p69yGszAOHp39?orgId=2

some preliminary notes:

1) kafka
* disk space saved: haven't found a good way to measure locally (we can just measure when we deploy to ops) but until then io (see below) should give an indication
* io saved is not as much as I expected:
  fakemetrics messages were 203B before, now they are 33B which is a significant reduction to only 16%
  However, IO on kafka and MT containers reduced by about half. maybe we underestimated the overhead kafka adds to each message
  (i believe it adds some metadata like a timestamp). We can probably get more significant savings by batching MetricPoint's per message.
* interestingly, kafka CPU also reduced by about 30%

2) MT
* cpu reduction in line with what was expected. 20~30%
* memory usage -> need to do more experiments
* MT allocation rate dropped by just over 50%. sweet. allocation profile is looking very clean.
  most allocations are now coming from our sarama library and the snappy decoding.
  (I wonder how that will change with the switch to confluent).
  also GC frequency reduced by over 50%

3) fakemetrics (representative of tsdb-gw)
* cpu is about the same 
* memory is reduced.  i had expected memory usage would be higher due to the keycache we need now. I guess the more efficient MetricPoint encoding overcompensates for it. (yeay)

Dieterbe commented:
here's a fresh stack start and 10 minute long 25kHz ingest, comparing old MT with old payloads, vs new MT with new payloads
old: https://snapshot.raintank.io/dashboard/snapshot/eEZv5VleN95vz4ovh3bA3IiKGtmXFWT3?orgId=2
new: https://snapshot.raintank.io/dashboard/snapshot/HImyLItr6jTw9AMnb22AigDmczADmdmF?orgId=2

conclusion:
* no significant change in heap objects. this is surprising to me. I think the main takeway here is that while we (I) thought live non-recyclable pointers was the big problem, over recycleable pointers (garbage), it seems all the recycleable objects that get allocated (as part of sarama/snappy, etc)  outweigh the live pointers. note that v2 is able to reach lower minima (probably due to reduced live pointers), the avg and max is higher. perhaps because GC is now less frequent allowing more garbage to pile up
* the GC stats confirm this I think: while collections are only half as frequent, and some of them were shorter, the avg and max is higher. though we know GC stop-the-world is a very misleading metric that in our case barely scratches the surface of conveying true "GC impact" (in the concurrent phases, which are not measured). GC Cpu fraction promille went down which is nice.
* MT RAM usage is reduced by about 1.7% ; I expected better. looking at a profile, turns out 83% of memory in use is held by `sarama.(*consumer).ConsumePartition`. curious to see how this will change with confluent.

Dieterbe commented:
blocked on https://github.com/grafana/metrictank/issues/260

Dieterbe commented:
## status update:

* benchmarks (see above) are satisfactory, though not entirely in line with expectations regarding kafka io/disk space.
* all tools build. had to remove mt-replicate-via-tsdb, will have to devise a new intra-cluster replication method since tsdb-gw doesn't support MetricPoint messages
* all unit tests pretty much work. there's some breakage in `idx` which is probably quickly resolved once #260 is fixed

## TODO
* code reviews
* some more functionality tests

Dieterbe commented:
for completeness, i wanted to know snappy vs compression=none
4 runs:
old format, comp none
new format, comp none
old format, snappy
new, snappy

https://snapshot.raintank.io/dashboard/snapshot/dghBYbu56QnVwSUWkp03jXZbCm5K1l42?orgId=2
as you can see there's nothing to gain by disabling compression:
kafka cpu went way up, network io went up (obviously), fakemetrics cpu about the same (maybe about a % gain), MT cpu the same.
I can't post the MT snapshot but all it shows is that in the very first run, MT built up a kafka lag (probably because kafka hit 100% cpu) and it took another minute or so after the real run for it to catch up

woodsaj commented on idx/idx.go:
```diff
@@ -99,39 +55,41 @@ type MetricIndex interface {
 	Stop()
 
+	// UpdateMaybe updates an existing archive, if found.
+	// it returns the existing archive (if any), and whether it was found
```
This comment describes 2 of the returned values. What is the other one?

woodsaj commented on idx/idx.go:
```diff
@@ -99,39 +55,41 @@ type MetricIndex interface {
 
+	// UpdateMaybe updates an existing archive, if found.
+	// it returns the existing archive (if any), and whether it was found
+	UpdateMaybe(point schema.MetricPoint, partition int32) (Archive, int32, bool)
```
can we just call this `Update()`   updateMaybe sounds weird to me.  It makes it sounds like the update may or may not happen due to some unknown reason.

This is just a simple update request that will either succeed or fail due to there being being no existing object to update.

DanCech replied:
+1 I'm not a fan of the BlahMaybe naming in general.

Dieterbe replied:
+1 i like it

woodsaj commented on idx/memory/memory.go:
```diff
@@ -206,18 +206,45 @@ func (m *MemoryIdx) Stop() {
+	existing, ok := m.defById[point.MKey]
+	if ok {
+		oldPart := existing.Partition
+		log.Debug("metricDef with id %v already in index", point.MKey)
```
we should probably get rid of log lines that fire on every datapoint received.

a single log.Debug call needs
```
BenchmarkLog-4   	 1000000	      1152 ns/op	      64 B/op	       2 allocs/op
```
https://gist.github.com/woodsaj/33eafd6976181797dfbb46f193f6bcc1

Dieterbe replied:
OTOH, that's what the debug log calls are for, so you can follow in high detail what happens to a metric at each step along the way. we've historically had these sort of calls in several places (also the input handlers and input plugins).
I agree - obviously - that we should minimize their overhead though. in a few places we guard expensive debug calls with `if LogLevel < 2`, but that's basically to address a shortcoming in our logging library. when we switch to logrus this should no longer be an issue at all (https://github.com/imkira/go-loggers-bench#textnegative-1)

anyway, logging performance tweaks are out of scope for this PR.

woodsaj replied:
after a quick look over the code, this is the only place we are calling log.Debug at an insane frequency.  Until we replace the logging lib, lets comment this log line out.  It will literally save us millions of allocations a second across our clusters.

Dieterbe replied:
it is not. (see kafkaMdm.consumePartition)
but anyway i addressed this in 45c02a0

woodsaj commented on input/input.go:
```diff
@@ -44,31 +46,59 @@ func NewDefaultHandler(metrics mdata.Metrics, metricIndex idx.MetricIndex, input
-func (in DefaultHandler) Process(metric *schema.MetricData, partition int32) {
-	if metric == nil {
+func (in DefaultHandler) ProcessMetricPoint(point schema.MetricPoint, partition int32) {
+	in.metricsReceived.Inc()
```
I think this should be a separate metric.  it would be useful to see the ratio of metricPoint:MetricData being ingested

Dieterbe replied:
right. will also do the same for the invalid metric

woodsaj commented on input/input.go:
```diff
@@ -44,31 +46,59 @@ func NewDefaultHandler(metrics mdata.Metrics, metricIndex idx.MetricIndex, input
+	archive, _, ok := in.metricIndex.UpdateMaybe(point, partition)
+	in.pressureIdx.Add(int(time.Since(pre).Nanoseconds()))
+
+	if !ok {
```
We should keep a metric of the number of points being dropped due to not existing in the index.

woodsaj commented on vendor/gopkg.in/raintank/schema.v1/archive.go:
```diff
@@ -0,0 +1,90 @@
+
+// important: caller must make sure to call IsSpanValid first
+func NewArchive(method Method, span uint32) Archive {
+	code := spanHumanToCode[span]
```
do we need to make sure that span is a supported value?

woodsaj replied:
nevermind, the comment makes it clear that is the callers responsibility

woodsaj commented on vendor/gopkg.in/raintank/schema.v1/metric.go:
```diff
@@ -113,7 +112,7 @@ type MetricDataArray []*MetricData
-	Id       string `json:"id"`
+	Id       MKey   `json:"id"`
 	OrgId    int    `json:"org_id"`
 	Name     string `json:"name" elastic:"type:string,index:not_analyzed"` // graphite format
```
lets clean the struct tags up and remove the "elastic" hints.

woodsaj commented on vendor/gopkg.in/raintank/schema.v1/metric.go:
```diff
@@ -113,7 +112,7 @@ type MetricDataArray []*MetricData
+	Id       MKey   `json:"id"`
 	OrgId    int    `json:"org_id"`
 	Name     string `json:"name" elastic:"type:string,index:not_analyzed"` // graphite format
 	Metric   string `json:"metric"`                                        // kairosdb format (like graphite, but not including some tags)
```
i would really like to get rid of "Metric" it is currently always set to the same value as "Name"

woodsaj replied:
We should also update SetId() to use m.Name instead of m.Metric. and remove the Validate() rule requiring it.
We can also safely drop the field from the metricData payload.  If clients continue to send it, it will just be ignored when marshaling the data.

Dieterbe replied:
@woodsaj and same for the index right? I'll update the code to remove the property from the structures and basically ignore that field in cassandra.

Dieterbe commented:
rebased on top of master, now we have all the org id stuff and related idx fixes
addressed all comments

===> time for another/last round of reviews ! and then let's merge and deploy ! :)
@woodsaj @replay @shanson7
